src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -151,11 +156,26 @@ public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper)
             }
         }
 
-        internal unsafe struct IUnknownVftbl
+        internal unsafe class NativeObjectWrapper
         {
-            public delegate* unmanaged<IntPtr, ref Guid, out IntPtr, int> QueryInterface;
-            public delegate* unmanaged<IntPtr, uint> AddRef;
-            public delegate* unmanaged<IntPtr, uint> Release;
+            private IntPtr _externalComObject;
+            private ComWrappers _comWrappers;
+            public GCHandle ProxyHandle;
+
+            public NativeObjectWrapper(IntPtr externalComObject, ComWrappers comWrappers, object comProxy)
+            {
+                _externalComObject = externalComObject;
+                _comWrappers = comWrappers;
+                ProxyHandle = GCHandle.Alloc(comProxy, GCHandleType.Weak);
+                Marshal.AddRef(externalComObject);
+            }
+
+            ~NativeObjectWrapper()
+            {
+                _comWrappers.RemoveRCWFromCache(_externalComObject);
+                _proxyHandle.Free();
```

Comment:
This can be unallocated handle if GCHandle.Alloc in the constructor fails. This should get the IsAllocated check.

Reply:
Done